kernel/csymtable.ml

```diff
@@ -26,65 +26,153 @@ open Cbytegen
+    try Int.Map.modify hash (fun _ l -> (k, v) :: l) m
+    with Not_found -> Int.Map.add hash [k, v] m
+  let find k m =
+    let hash = M.hash k in
+    let l = Int.Map.find hash m in
+    List.assoc_f M.equal k l
+end
+
+module SConstTable = HashMap (struct
   type t = structured_constant
   let equal = eq_structured_constant
   let hash = hash_structured_constant
 end)
 
-module AnnotTable = Hashtbl.Make (struct
+module AnnotTable = HashMap (struct
   type t = annot_switch
   let equal = eq_annot_switch
   let hash = hash_annot_switch
 end)
 
-module ProjNameTable = Hashtbl.Make (Projection.Repr)
+module ProjNameTable = HashMap (Projection.Repr)
+
+module GlobVal :
+sig
+  type key = int
+  type t
+  val empty : int -> t
+  val push : t -> values -> key * t
+  val vm_global : t -> vm_global
+  (** Warning: due to sharing, the resulting value can be modified by
+      persistent operations. When calling this function one must ensure
+      that no other context modification is performed before the value drops
+      out of scope. *)
+end =
+struct
+  type key = int
+
+  type view =
+  | Root of values array
+  | DSet of int * values * view ref
+
+  type t = {
+    data : view ref;
+    size : int;
+    (** number of actual elements in the array *)
+  }
+
+  let empty n = {
+    data = ref (Root (Array.make n crazy_val));
+    size = 0;
+  }
+
+  let rec rerootk t k = match !t with
+  | Root _ -> k ()
+  | DSet (i, v, t') ->
+    let next () = match !t' with
+    | Root a as n ->
+      let v' = Array.unsafe_get a i in
+      let () = Array.unsafe_set a i v in
+      let () = t := n in
+      let () = t' := DSet (i, v', t) in
+      k ()
+    | DSet _ -> assert false
+    in
+    rerootk t' next
+
+  let reroot t = rerootk t (fun () -> ())
+
+  let push { data = t; size = i } v =
+    let () = reroot t in
+    match !t with
+    | DSet _ -> assert false
+    | Root a as n ->
+      let len = Array.length a in
+      let data =
+        if i < len then
+          let old = Array.unsafe_get a i in
+          if old == v then t
+          else
+            let () = Array.unsafe_set a i v in
+            let res = ref n in
+            let () = t := DSet (i, old, res) in
+            res
+        else
+          let nlen = 2 * len + 1 in
+          let nlen = min nlen Sys.max_array_length in
+          let () = assert (i < nlen) in
+          let a' = Array.make nlen crazy_val in
+          let () = Array.blit a 0 a' 0 len in
+          let () = Array.unsafe_set a' i v in
+          let res = ref (Root a') in
+          let () = t := DSet (i, crazy_val, res) in
+          res
+      in
+      (i, { data; size = i + 1 })
+
+  let vm_global { data; _ } =
+    let () = reroot data in
+    match !data with
+    | DSet _ -> assert false
+    | Root a -> (a : vm_global)
+
+end
+
 
-let str_cst_tbl : int SConstTable.t = SConstTable.create 31
+(*******************)
+(* Linkage du code *)
+(*******************)
+
+(* Table des globaux *)
```

Comment:
French

Reply:
This is merely moving code, I didn't introduce this comment.